sklearn/metrics/cluster/_expected_mutual_info_fast.pyx

```diff
@@ -7,23 +7,25 @@ from scipy.special import gammaln
 import numpy as np
 cimport numpy as cnp
 
-cnp.import_array()
-ctypedef cnp.float64_t DOUBLE
 
-
-def expected_mutual_information(contingency, int n_samples):
+def expected_mutual_information(contingency, cnp.int64_t n_samples):
     """Calculate the expected mutual information for two labelings."""
-    cdef int R, C
-    cdef DOUBLE N, gln_N, emi, term2, term3, gln
-    cdef cnp.ndarray[DOUBLE] gln_a, gln_b, gln_Na, gln_Nb, gln_nij, log_Nnij
-    cdef cnp.ndarray[DOUBLE] nijs, term1
-    cdef cnp.ndarray[DOUBLE] log_a, log_b
-    cdef cnp.ndarray[cnp.int32_t] a, b
-    #cdef np.ndarray[int, ndim=2] start, end
-    R, C = contingency.shape
-    N = <DOUBLE>n_samples
-    a = np.ravel(contingency.sum(axis=1).astype(np.int32, copy=False))
-    b = np.ravel(contingency.sum(axis=0).astype(np.int32, copy=False))
+    cdef:
+        cnp.float64_t emi = 0
+        cnp.int64_t n_rows, n_cols
+        cnp.float64_t term2, term3, gln
+        cnp.int64_t[::1] a_view, b_view
+        cnp.float64_t[::1] nijs_view, term1
+        cnp.float64_t[::1] gln_a, gln_b, gln_Na, gln_Nb, gln_Nnij, log_Nnij
+        cnp.float64_t[::1] log_a, log_b
+        Py_ssize_t i, j, nij
+        cnp.int64_t start, end
+
+    n_rows, n_cols = contingency.shape
+    a = np.ravel(contingency.sum(axis=1).astype(np.int64, copy=False))
+    b = np.ravel(contingency.sum(axis=0).astype(np.int64, copy=False))
```

Comment:
I assume this is here where we could avoid the copy if the contingency matrix was already passed as a 64 bits array.